Sources/Server/main.swift

```diff
@@ -0,0 +1,36 @@
+import Dependencies
+import HttpPipeline
+import NIO
+import PointFree
+import Prelude
+import Dispatch
+
+// NB: this DispachGroup/Task is necessary to work around strange deadlock issues in Swift async.
+//     For more info, see:
+//     https://github.com/apple/swift-nio/blob/1ce136b4c392bd7427cf408d65305a05d162fb46/Sources/NIOAsyncAwaitDemo/main.swift#L66-L76.
+let group = DispatchGroup()
+group.enter()
+Task {
```

Comment:
@weissi Sorry to ping you out of the blue, but we found this trick from your PR https://github.com/apple/swift-nio/pull/1701, but there wasn't much info about why it's needed or why it works.

Without this we were seeing very strange behavior in processing requests. A request would come into the server, and our `siteMiddleware` would finish processing it (even the `defer` would execute inside), yet somehow the function was never fully exited. It would just hang forever. It honestly seemed like a Swift language bug.

This `DispatchGroup`/`Task` trick fixed it for us. Is it safe to use in production?

Reply:
These days you can use async main functions and that's best.

Reply:
If you can't do that, then the best workaround is probably something along the lines of

```
Task {
    try await stuff()
    DispatchQueue.main.async {
        exit(0)
    }
}

RunLoop.main.run()
```

Reply:
> These days you can use async main functions and that's best.

Yeah, that's exactly how we started, but we had very weird issues with it. This task group trick was the thing that fixed it.

But now I've just found another workaround to fix it, which is to pass `siteMiddleware` directly to our `run` helper instead of expanding it into a closure:

```diff
// Server
 run(
-  { await siteMiddleware($0) },
+  siteMiddleware,
   on: envVars.port,
```

This has got to be a Swift concurrency bug.

Reply:
Thanks for the info @weissi. We're going to try to get a minimal repro to submit a bug report.